tokio-trace-macros/tokio-trace-macros/src/lib.rs

```diff
@@ -0,0 +1,47 @@
+#![feature(proc_macro_diagnostic)]
+
+extern crate proc_macro;
+#[macro_use]
+extern crate syn;
+#[macro_use]
+extern crate quote;
+extern crate proc_macro2;
+
+use proc_macro::TokenStream;
+use proc_macro2::Span;
+use syn::token::{Async, Const, Unsafe};
+use syn::{Abi, Attribute, Block, Ident, ItemFn, Visibility};
+
+#[proc_macro_attribute]
+pub fn trace(_args: TokenStream, item: TokenStream) -> TokenStream {
+    let input: ItemFn = parse_macro_input!(item as ItemFn);
```

Comment:
Also for potential future work: it would be cool to allow things that aren't functions to be traced. At the very least, it would be nice to detect if we're annotating a future, and instrument it in the "right way" --- by constructing the span once and wrapping the future in a `tokio_trace_futures::Instrumented` future. 

The naïve approach --- just putting the `#[trace]` macro on the implementation of `Future::poll` for that future --- _wouldn't_ do the Right Thing; it would instead create a _new_ span that's entered for every individual time that future is polled. This could potentially be a big enough gotcha that I think it's probably worthwhile to special-case this.

We'd probably want to figure out a similar approach for `async fn`s as well...

Reply:
> Also for potential future work: it would be cool to allow things that aren't functions to be traced. At the very least, it would be nice to detect if we're annotating a future, and instrument it in the "right way" --- by constructing the span once and wrapping the future in a tokio_trace_futures::Instrumented future.

> The naïve approach --- just putting the #[trace] macro on the implementation of Future::poll for that future --- wouldn't do the Right Thing; it would instead create a new span that's entered for every individual time that future is polled. This could potentially be a big enough gotcha that I think it's probably worthwhile to special-case this.

That's a good point, and I'd need to dig deeper into the Syn docs.

> We'd probably want to figure out a similar approach for async fns as well...

It already is! See the `#asyncness` [on line 41](https://github.com/hawkw/tokio-trace-prototype/blob/6a1c91edbe6795a4f5c99f999e1406fef4234cdc/tokio-trace-macros/tokio-trace-macros/src/lib.rs#L41)? That's all the work we need to do to support async functions, from a macro standpoint.

Reply:
> > We'd probably want to figure out a similar approach for async fns as well...
>
> It already is! See the #asyncness on line 41? That's all the work we need to do to support async functions, from a macro standpoint.

Yes, but I think we'd want to make sure we do the _right thing_ for `async fn`s --- namely, I think we'd want to make sure that the _future created by that async fn_ is instrumented by that span, rather than just the initial call to the `async fn` that returns the future. This may require some testing.